### PR TITLE
apply events late

### PR DESCRIPTION
**Pull request by prolic**
it applies events to aggregate roots only after they are written to the event stream

### CODE REVIEW

File: src/Aggregate/ConfigurableAggregateTranslator.php [hunk 1]
```diff
@@ -72,6 +79,11 @@ public function __construct(
             $this->popRecordedEventsMethodName = $popRecordedEventsMethodName;
         }
 
+        if (null !== $applyRecordedEventsMethodsName) {
+            Assertion::minLength($applyRecordedEventsMethodsName, 1, 'Pop recorded events method name needs to be a non empty string');
```
Comment (codeliner) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: s/Pop recorded/Apply recorded

File: src/Aggregate/ConfigurableAggregateTranslator.php [hunk 2]
```diff
@@ -211,4 +223,37 @@ public function extractPendingStreamEvents($eventSourcedAggregateRoot)
...
+
+        foreach ($events as $event) {
+            if (!$event instanceof Message) {
+                throw new AggregateTranslationFailedException(sprintf(
+                    'A recorded event of the aggregate root %s has the wrong type. Expected Prooph\Common\Messaging\Message. Got %s',
+                    get_class($eventSourcedAggregateRoot),
+                    is_object($event)? get_class($event) : gettype($event)
+                ));
+            }
+            $eventSourcedAggregateRoot->{$this->applyRecordedEventsMethodName}($event);
```
Comment (codeliner) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: Before applying the event we need to check if a `messageToEventCallback` was injected in the translator. If so the callback needs to be invoked first.

**Comment by codeliner** [comment]
@prolic I love this change. It is the best way to solve the consistency problem :+1: It is a BC break but I cannot think of better solution without breaking BC.
Two minor issues (see inline comments) and we need to align the docs. If you want, you can create a follow up ticket for documentation and I take it over.

**Comment by prolic** [comment]
done.

**Pull request merged**